src/sysroots.c

```diff
@@ -0,0 +1,230 @@
+    Char *       q;
+    Int          i;
+    Int          n;
+
+    /* set string to a default value if unset                              */
+    if (string == 0 || *string == 0) {
+        string = "./";
+    }
+
+    /*
+    ** check if we append, prepend or overwrite.
+    */
+    if (string[0] == ';') {
+        /* Count the number of root directories already present.           */
+        n = 0;
+        while (SyGapRootPaths[n][0] != '\0')
+            n++;
+
+        /* Skip leading semicolon.                                        */
+        string++;
+    }
+    else if (string[strlen(string) - 1] == ';') {
+        /* Count the number of directories in 'string'.                    */
+        n = 0;
+        p = string;
+        while (*p)
+            if (*p++ == ';')
+                n++;
+
+        /* Find last root path.                                            */
+        for (i = 0; i < MAX_GAP_DIRS; i++)
+            if (SyGapRootPaths[i][0] == '\0')
+                break;
+        i--;
+
+#ifdef HPCGAP
+        n *= 2;    // for each root <ROOT> we also add <ROOT/hpcgap> as a root
+#endif
+
+        /* Move existing root paths to the back                            */
+        if (i + n >= MAX_GAP_DIRS)
+            return;
+        while (i >= 0) {
+            memcpy(SyGapRootPaths[i + n], SyGapRootPaths[i],
+                   sizeof(SyGapRootPaths[i + n]));
+            i--;
+        }
+
+        n = 0;
+    }
+    else {
+        /* Make sure to wipe out all possibly existing root paths          */
+        for (i = 0; i < MAX_GAP_DIRS; i++)
+            SyGapRootPaths[i][0] = '\0';
+        n = 0;
+    }
+
+    /* unpack the argument                                                 */
+    p = string;
+    while (*p) {
+        if (n >= MAX_GAP_DIRS)
+            return;
+
+        q = SyGapRootPaths[n];
+        while (*p && *p != ';') {
+            *q = *p++;
+
+#ifdef SYS_IS_CYGWIN32
+            // change backslash to slash for Windows
+            if (*q == '\\')
+                *q = '/';
+#endif
+
+            q++;
+        }
+        if (q == SyGapRootPaths[n]) {
+            strxcpy(SyGapRootPaths[n], "./", sizeof(SyGapRootPaths[n]));
+        }
+        else if (q[-1] != '/') {
+            *q++ = '/';
+            *q = '\0';
+        }
+        else {
+            *q = '\0';
+        }
+        if (*p) {
+            p++;
+        }
+        n++;
+#ifdef HPCGAP
+        // for each root <ROOT> to be added, we first add <ROOT/hpcgap> as a root
+        if (n < MAX_GAP_DIRS) {
+            strlcpy(SyGapRootPaths[n], SyGapRootPaths[n - 1],
+                    sizeof(SyGapRootPaths[n]));
+        }
+        strxcat(SyGapRootPaths[n - 1], "hpcgap/",
+                sizeof(SyGapRootPaths[n - 1]));
+        n++;
+#endif
+    }
+
+    // replace leading tilde ~ by HOME environment variable
+    // TODO; instead of iterating over all entries each time, just
+    // do this for the new entries
```

Comment:
FYI I left this code this way as it was the minimal change to achieve the goal, with least risk of introducing a regression.

Overall it would be nicer to store all these paths as GAP objects (strings and plain lists), but GASMAN is not yet initialized when this function is called... because it is called when we parse command line options, and those include settings for GASMAN... one solution to that would be to parse command line options twice, once for GASMAN options, then again for everything else...